### ISSUE TITLE
Overflow of section_size possible in coda-grib.c

### ISSUE THREAD

**Issue by schwehr** [stcorp/coda#46]
```
section_size = ((buffer[0] * 256 + buffer[1]) * 256 + buffer[2]) * 256 + buffer[3];
```

Gives:

```
third_party/stcorp_coda/fuzz/coda_recognize_file_fuzzer crash-af51b1fb3ec5c81523e08e7e7c0a567c34650366
Running the target on file crash-af51b1fb3ec5c81523e08e7e7c0a567c34650366 (97 bytes)
third_party/stcorp_coda/libcoda/coda-grib.c:2216:59: runtime error: signed integer overflow: 16777215 * 256 cannot be represented in type 'int'
    #0 0x559fc57e6896 in read_grib2_message third_party/stcorp_coda/libcoda/coda-grib.c:2216:59
    #1 0x559fc57d3e0f in coda_grib_reopen third_party/stcorp_coda/libcoda/coda-grib.c:3138:17
    #2 0x559fc5807eb6 in reopen_with_backend third_party/stcorp_coda/libcoda/coda-product.c:408:17
    #3 0x559fc5805a5d in open_file third_party/stcorp_coda/libcoda/coda-product.c:550:9
    #4 0x559fc58054dc in coda_recognize_file third_party/stcorp_coda/libcoda/coda-product.c:594:9
    #5 0x559fc576024c in LLVMFuzzerTestOneInput third_party/stcorp_coda/fuzz/coda_recognize_file_fuzzer.cc:16:3
```

Proposed solution:

```c
#include <limits.h>

// SNIP

    uint64_t section_size_tmp =
        ((buffer[0] * 256 + buffer[1]) * 256 + buffer[2]) * 256 + buffer[3];
    if (section_size_tmp > UINT_MAX)
    {
        return -1;
    }
    section_size = section_size_tmp;
```

See also:
- https://wiki.sei.cmu.edu/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow
- https://wiki.sei.cmu.edu/confluence/display/c/INT30-C.+Ensure+that+unsigned+integer+operations+do+not+wrap

**Comment by schwehr** [comment]
[crash-af51b1fb3ec5c81523e08e7e7c0a567c34650366.zip](https://github.com/stcorp/coda/files/4468071/crash-af51b1fb3ec5c81523e08e7e7c0a567c34650366.zip)

**Comment by schwehr** [comment]
Investigating another crash, I see that my fix above isn't good enough.  Making all the constants be uint64 should force all the intermediate values to be large enough.

```c
    uint64_t section_size_tmp =
        ((buffer[0] * 256ul + buffer[1]) * 256ul + buffer[2]) * 256ul + buffer[3];
    if (section_size_tmp > UINT_MAX)
    {
        return -1;
    }
    section_size = section_size_tmp;
```

**Comment by rouault** [comment]
@schwehr Your fix is suboptimal. There's no way that reading 4 bytes to make a uint32 can overflow it... You just need to cast to uint32 the highest byte.
```
uint32_t section_size=
        ((uint32_t)buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];
```
An optimizing compiler generates the bswap instruction for this on Intel arch

**Comment by svniemeijer** [comment]
The solution of @rouault is indeed the correct one.
Fixed in fbb38868bb5b6d16bb7b21c2871ed2f757cc4827

**Comment by schwehr** [comment]
Yes.  Most of my fixes are just the first thing I see to get the process going.   I am exhausted and trying squeeZe work hours when the kids let me. There are no tests in the project, so it's hard to know if something breaks other projects dependent on specific behavior.
